File: java/src/main/java/io/concord/wordcounter/SentenceSplitter.java
package io.concord.wordcounter;

import io.concord.*;
import io.concord.swift.*;
import java.util.*;

public class SentenceSplitter extends Computation {
  
  public void init(ComputationContext ctx) {
    System.out.println("SentenceSplitter.java initialized");
  }

  public void destroy() {
    System.out.println("SentenceSplitter.java destructing");
  }

  public void processTimer(ComputationContext ctx, String key, long time) {
    throw new RuntimeException("Method not implemented");
  }
  
  public void processRecord(ComputationContext ctx, Record record) {
    try {
      String sentences = new String(record.getKey(), "UTF-8");    
      String words[] = sentences.split(" ");
      for (int i = 0; i < words.length; ++i) {
	ctx.produceRecord("words".getBytes(), words[i].getBytes(), "-".getBytes());
      }
    } catch(Exception e) {
      throw new RuntimeException(e);
    }
  }

  public Metadata metadata() {
    Set<StreamTuple> istreams = new HashSet<StreamTuple>();
    istreams.add(new StreamTuple("sentences", StreamGrouping.SHUFFLE));
    return new Metadata("sentence-splitter",
			istreams,
			new HashSet<String>(Arrays.asList("words")));
  }
  
  public static void main(String[] args) {
    ServeComputation.serve(new SentenceSplitter());
  }

}

File: scala/src/main/scala/io/concord/wordcounter/SentenceSplitter.scala
package io.concord.wordcounter

import java.util.{ HashSet => MutableHashSet }

import io.concord._
import io.concord.swift._

class SentenceSplitter extends Computation {
  override def init(ctx: ComputationContext): Unit = {
    println(s"${this.getClass.getSimpleName} initialzed")
  }

  override def destroy(): Unit = {
    println(s"${this.getClass.getSimpleName} destructing")
  }

  override def processRecord(ctx: ComputationContext, record: Record): Unit = {
    val sentences = new String(record.getKey, "UTF-8")
    val words = sentences.split(" ")
    words.foreach((word) => {
      ctx.produceRecord("words".getBytes, word.getBytes, "-".getBytes)
    })
  }

  override def processTimer(ctx: ComputationContext, key: String, time: Long): Unit = ???

  override def metadata(): Metadata = {
    val ostreams = new MutableHashSet[String](java.util.Arrays.asList("words"))
    val istreams = new MutableHashSet[StreamTuple]()
    istreams.add(new StreamTuple("sentences", StreamGrouping.SHUFFLE))
    new Metadata("sentence-splitter", istreams, ostreams)
  }

}

object SentenceSplitter {
  def main(args: Array[String]): Unit = {
    ServeComputation.serve(new SentenceSplitter())
  }
}
